Clarify description of tracestate value (#411)
w3c/distributed-tracing: spec/20-http_request_header_format.md

##### Value

> The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This effectively means that all leading spaces MUST be preserved as part of the value and all trailing spaces MUST not be preserved.

Reviewer: It is still allowed to preserve the OWS characters.

```suggestion
The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This effectively means that all leading spaces MUST be preserved as part of the value and all trailing spaces MAY not be preserved.
```

Author: Really? The grammar ends with nblk-chr. Doesn't that mean it's not ok to preserve trailing space?

Reviewer: There is OWS allowed between kv pairs

```
key1=value1 , key2=value2
           ^this is optional whitespace, not part of the value, and may be preserved or dropped
```

Author: Ah maybe we're not on the same page on this

> not part of the value, and may be preserved or dropped

What did it mean to be not part of the value, but preserved? Does it mean that the value MUST be presented to user space (i.e., SDK) without trailing space but when propagating it's ok to keep or drop it? If so it makes sense and I'll add the caveat about propagation.

Reviewer: Maybe this
```suggestion
The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value.
```

Author: Thanks this makes sense but is still a bit unclear about what it means, in practice, to be "considered to be optional whitespace". I added a point about propagation which is my understanding but I'm not sure. It's why I want to make sure this spec is clear :)

> The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. They MAY be excluded when propagating the header.

Reviewer: I think "they" isn't entirely clear

```suggestion
The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. Optional whitespace MAY be excluded when propagating the header.
```

Author: Yup makes sense thanks

> The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. Optional whitespace MAY be excluded when propagating the header.

Reviewer 2: simplify sentences.
```suggestion
The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=). The string must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. All leading spaces MUST be preserved as part of the value. All trailing spaces are considered to be optional whitespace characters not part of the value. Optional trailing whitespace MAY be excluded when propagating the header.
```

Reviewer 2: > Note that this also excludes tabs, newlines, carriage returns, etc.

It's not clear what this is referring to - the whole value (i.e. "printable ASCII") or just the trailing characters. Could we not prohibit non-0x20 whitespace completely, both leading and trailing?

> The value is an opaque string containing up to 256 printable ASCII [RFC0020] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. Optional whitespace MAY be excluded when propagating the header.

Reviewer 3: ```suggestion
The value is an opaque string containing up to 256 printable ASCII [[!RFC0020]] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. Optional whitespace MAY be excluded when propagating the header.
```

This `[[!RFC0020]]` link does indeed look odd on Github and might look like a typo but it's actually used on purpose throughout the document. In the document published by W3C, this adds the referenced RFC to the bibliography on the bottom and adds a link to that: https://www.w3.org/TR/trace-context/#value

Reviewer: ah good catch I missed the link being broken

Author: Thanks! Sorry think it was lost in a copy-paste

> The value is an opaque string containing up to 256 printable ASCII [[!RFC0020]] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=) which must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. This means that all leading spaces MUST be preserved as part of the value and all trailing spaces are considered to be optional whitespace characters not part of the value. Optional whitespace MAY be excluded when propagating the header.

Author: ```suggestion
The value is an opaque string containing up to 256 printable ASCII [[!RFC0020]] characters (i.e., the range 0x20 to 0x7E) except comma (,) and (=). The string must end with a character which is not a space (0x20). Note that this also excludes tabs, newlines, carriage returns, etc. All leading spaces MUST be preserved as part of the value. All trailing spaces are considered to be optional whitespace characters not part of the value. Optional trailing whitespace MAY be excluded when propagating the header.
```
